File: python3/FindAllAnagrams.py
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        pHash = [0] * 26
        
        for char in p:           
            pHash[ord(char) - ord('a')] += 1
                
        curStart = 0
        curIndex = 0
        output = []
        curHash = [0] * 26
        
        while (curIndex < len(s)):
            curHash[ord(s[curIndex]) - ord('a')] += 1
            curIndex += 1      
            if (curIndex - curStart) == len(p):
                if (curHash == pHash):
                    output.append(curStart)
                
                curHash[ord(s[curStart]) - ord('a')] -= 1
                curStart += 1
                
        return output

File: c++/FindAllAnagrams.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int curStart = 0;
        int curIndex = 0;
        
        vector<int> toFind(26, 0);
        vector<int> found(26, 0);
        vector<int> output;
        
        for (auto c : p) {
            toFind[c -'a'] += 1;
        }
        
        while (curIndex < s.length()) {
            found[s[curIndex] - 'a'] += 1;
            ++curIndex;
            
            if ((curIndex-curStart) == p.length()) {
                if (toFind == found) {
                    output.push_back(curStart);
                }
                
                found[s[curStart] - 'a'] -= 1;
                curStart += 1;
            }
        }
        
        return output;
    }
};
